File: tcu_validation_framework/src/reporting/ReportGenerator.cpp
/**
 * @file ReportGenerator.cpp
 * @brief HTML/JSON/CSV test report generation.
 */

#include "reporting/ReportGenerator.h"
#include "logging/Logger.h"

#include <fstream>
#include <iomanip>
#include <sstream>
#include <ctime>
#include <filesystem>
#include <nlohmann/json.hpp>

namespace tcu::reporting {

using json = nlohmann::json;
static auto s_log = tcu::logging::Logger::get("reporting");

// ============================================================
// Construction
// ============================================================

ReportGenerator::ReportGenerator(const std::string& output_dir)
    : m_output_dir(output_dir)
{
    std::filesystem::create_directories(output_dir);
}

// ============================================================
// Main generate
// ============================================================

bool ReportGenerator::generate(const tcu::validation::SuiteResult& suite,
                                ReportFormat fmt,
                                const std::string& filename_prefix) {
    bool ok = true;
    if (fmt == ReportFormat::HTML || fmt == ReportFormat::ALL) {
        ok &= generate_html(suite, filename_prefix);
    }
    if (fmt == ReportFormat::JSON || fmt == ReportFormat::ALL) {
        ok &= generate_json(suite, filename_prefix);
    }
    if (fmt == ReportFormat::CSV || fmt == ReportFormat::ALL) {
        ok &= generate_csv(suite, filename_prefix);
    }
    return ok;
}

// ============================================================
// HTML Report
// ============================================================

bool ReportGenerator::generate_html(const tcu::validation::SuiteResult& suite,
                                     const std::string& prefix) {
    std::string path = m_output_dir + "/" + prefix + ".html";
    std::ofstream f(path);
    if (!f.is_open()) {
        s_log->error("Cannot open HTML output: {}", path);
        return false;
    }

    float pass_pct = suite.total_tests > 0
                     ? 100.0f * suite.passed / suite.total_tests
                     : 0.0f;

    f << R"(<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>TCU Validation Report — )" << suite.suite_name << R"(</title>
<style>
  :root { --pass:#22c55e; --fail:#ef4444; --skip:#f59e0b; --to:#8b5cf6; --err:#6b7280; }
  body  { font-family:'Segoe UI',Arial,sans-serif; background:#0f172a; color:#e2e8f0; margin:0; padding:20px; }
  h1    { color:#38bdf8; border-bottom:2px solid #1e3a5f; padding-bottom:10px; }
  .summary { display:flex; gap:16px; flex-wrap:wrap; margin:20px 0; }
  .card  { background:#1e293b; border-radius:8px; padding:16px 24px; min-width:110px; text-align:center; }
  .card .num { font-size:2em; font-weight:700; }
  .card .lbl { font-size:.8em; color:#94a3b8; text-transform:uppercase; letter-spacing:.05em; }
  .pass { color:var(--pass); } .fail { color:var(--fail); } .skip { color:var(--skip); }
  .timeout { color:var(--to); } .error { color:var(--err); }
  table { width:100%; border-collapse:collapse; margin-top:20px; }
  th    { background:#1e3a5f; color:#bae6fd; padding:10px; text-align:left; }
  td    { padding:10px; border-bottom:1px solid #1e293b; }
  tr:hover td { background:#1e293b; }
  .badge { border-radius:4px; padding:2px 8px; font-size:.8em; font-weight:600; }
  .badge-pass    { background:#14532d; color:var(--pass); }
  .badge-fail    { background:#450a0a; color:var(--fail); }
  .badge-skip    { background:#451a03; color:var(--skip); }
  .badge-timeout { background:#2e1065; color:var(--to); }
  .badge-error   { background:#1f2937; color:var(--err); }
  .progress-bar  { background:#1e293b; border-radius:4px; height:12px; overflow:hidden; margin-top:8px; }
  .progress-fill { height:100%; background:var(--pass); transition:width .3s; }
</style>
</head>
<body>
<h1>TCU Validation Framework — Test Report</h1>
<p>Suite: <b>)" << suite.suite_name << R"(</b> &nbsp;|&nbsp; Duration: <b>)" << suite.total_ms
       << R"( ms</b></p>
<div class="progress-bar"><div class="progress-fill" style="width:)" << static_cast<int>(pass_pct)
       << R"(%;"></div></div>
<div class="summary">
  <div class="card"><div class="num">)" << suite.total_tests << R"(</div><div class="lbl">Total</div></div>
  <div class="card pass"><div class="num">)" << suite.passed << R"(</div><div class="lbl">Pass</div></div>
  <div class="card fail"><div class="num">)" << suite.failed << R"(</div><div class="lbl">Fail</div></div>
  <div class="card skip"><div class="num">)" << suite.skipped << R"(</div><div class="lbl">Skip</div></div>
  <div class="card timeout"><div class="num">)" << suite.timed_out << R"(</div><div class="lbl">Timeout</div></div>
  <div class="card error"><div class="num">)" << suite.errored << R"(</div><div class="lbl">Error</div></div>
</div>
<table>
<thead><tr>
  <th>#</th><th>Test ID</th><th>Test Name</th><th>Verdict</th>
  <th>Duration (ms)</th><th>Attempt</th><th>Message</th>
</tr></thead>
<tbody>
)";

    int row = 0;
    for (const auto& r : suite.results) {
        ++row;
        std::string badge, verdict_str;
        switch (r.verdict) {
            case tcu::validation::Verdict::PASS:
                badge = "badge-pass"; verdict_str = "PASS"; break;
            case tcu::validation::Verdict::FAIL:
                badge = "badge-fail"; verdict_str = "FAIL"; break;
            case tcu::validation::Verdict::SKIP:
                badge = "badge-skip"; verdict_str = "SKIP"; break;
            case tcu::validation::Verdict::TIMEOUT:
                badge = "badge-timeout"; verdict_str = "TIMEOUT"; break;
            default:
                badge = "badge-error"; verdict_str = "ERROR"; break;
        }
        f << "<tr>\n"
          << "  <td>" << row << "</td>\n"
          << "  <td><code>" << r.test_id << "</code></td>\n"
          << "  <td>" << r.test_name << "</td>\n"
          << "  <td><span class=\"badge " << badge << "\">" << verdict_str << "</span></td>\n"
          << "  <td>" << r.duration_ms << "</td>\n"
          << "  <td>" << r.attempt << "</td>\n"
          << "  <td>" << r.message << "</td>\n"
          << "</tr>\n";
    }

    f << R"(</tbody></table>
<p style="color:#475569;font-size:.8em;margin-top:24px;">
  Generated by TCU Validation Framework v2.0.0
</p>
</body>
</html>)";

    s_log->info("HTML report: {}", path);
    return true;
}

// ============================================================
// JSON Report
// ============================================================

bool ReportGenerator::generate_json(const tcu::validation::SuiteResult& suite,
                                     const std::string& prefix) {
    std::string path = m_output_dir + "/" + prefix + ".json";
    std::ofstream f(path);
    if (!f.is_open()) {
        s_log->error("Cannot open JSON output: {}", path);
        return false;
    }

    json doc;
    doc["suite_name"]  = suite.suite_name;
    doc["total_tests"] = suite.total_tests;
    doc["passed"]      = suite.passed;
    doc["failed"]      = suite.failed;
    doc["skipped"]     = suite.skipped;
    doc["timed_out"]   = suite.timed_out;
    doc["errored"]     = suite.errored;
    doc["total_ms"]    = suite.total_ms;
    doc["all_passed"]  = suite.all_passed;

    json results = json::array();
    for (const auto& r : suite.results) {
        json item;
        item["test_id"]     = r.test_id;
        item["test_name"]   = r.test_name;
        item["verdict"]     = static_cast<int>(r.verdict);
        item["duration_ms"] = r.duration_ms;
        item["attempt"]     = r.attempt;
        item["message"]     = r.message;
        results.push_back(item);
    }
    doc["results"] = results;

    f << doc.dump(2);
    s_log->info("JSON report: {}", path);
    return true;
}

// ============================================================
// CSV Report
// ============================================================

bool ReportGenerator::generate_csv(const tcu::validation::SuiteResult& suite,
                                    const std::string& prefix) {
    std::string path = m_output_dir + "/" + prefix + ".csv";
    std::ofstream f(path);
    if (!f.is_open()) {
        s_log->error("Cannot open CSV output: {}", path);
        return false;
    }

    // Header
    f << "test_id,test_name,verdict,duration_ms,attempt,message\n";

    auto escape_csv = [](const std::string& s) -> std::string {
        if (s.find_first_of(",\"\n\r") == std::string::npos) { return s; }
        std::string out = "\"";
        for (char c : s) {
            if (c == '"') { out += '"'; }
            out += c;
        }
        out += '"';
        return out;
    };

    auto verdict_to_str = [](tcu::validation::Verdict v) -> std::string {
        switch (v) {
            case tcu::validation::Verdict::PASS:    return "PASS";
            case tcu::validation::Verdict::FAIL:    return "FAIL";
            case tcu::validation::Verdict::SKIP:    return "SKIP";
            case tcu::validation::Verdict::TIMEOUT: return "TIMEOUT";
            default:                                return "ERROR";
        }
    };

    for (const auto& r : suite.results) {
        f << escape_csv(r.test_id)   << ","
          << escape_csv(r.test_name) << ","
          << verdict_to_str(r.verdict) << ","
          << r.duration_ms           << ","
          << r.attempt               << ","
          << escape_csv(r.message)   << "\n";
    }

    s_log->info("CSV report: {}", path);
    return true;
}

} // namespace tcu::reporting

File: adas_framework/utilities/report_generator.py
# adas_framework/utilities/report_generator.py
"""
Test Report Generator — ADAS Enterprise Framework.

Generates:
    1. Allure-compatible JSON fragments (allure serve / allure generate)
    2. HTML single-file summary (no external dependencies)
    3. Excel XLSX workbook with per-feature sheets and colour coding
    4. Grafana/InfluxDB metrics push (optional)

Usage:
    generator = ReportGenerator(cfg.report)
    generator.add_result(result)
    generator.finalize()        # Writes all formats
    generator.open_html()       # Opens browser preview
"""
from __future__ import annotations

import json
import os
import time
import html as _html
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional

try:
    from openpyxl import Workbook
    from openpyxl.styles import PatternFill, Font, Alignment
    _OPENPYXL = True
except ImportError:
    _OPENPYXL = False

from core.config import ReportConfig
from core.logger import report_log as log


# ─────────────────────────────────────────────────────────────────────────────
# Data model
# ─────────────────────────────────────────────────────────────────────────────

@dataclass
class TestResult:
    test_id:    str
    feature:    str
    status:     str           # "PASS" | "FAIL" | "SKIP" | "ERROR"
    duration_ms: float
    asil:       str           = "QM"
    req_ids:    List[str]     = field(default_factory=list)
    dtc_ids:    List[str]     = field(default_factory=list)
    error_msg:  str           = ""
    timestamp:  float         = field(default_factory=time.time)
    build:      str           = ""
    sw_version: str           = ""
    environment: str          = ""

    @property
    def passed(self) -> bool:
        return self.status == "PASS"

    @property
    def emoji(self) -> str:
        return {"PASS": "✅", "FAIL": "❌", "SKIP": "⚠️", "ERROR": "🔴"}.get(
            self.status, "❓"
        )


# ─────────────────────────────────────────────────────────────────────────────
# Colours
# ─────────────────────────────────────────────────────────────────────────────

_COLOR = {
    "PASS":  "#00C851",
    "FAIL":  "#FF4444",
    "SKIP":  "#FFBB33",
    "ERROR": "#CC0000",
}

_EXCEL_COLOR = {
    "PASS":  "00C851",
    "FAIL":  "FF4444",
    "SKIP":  "FFBB33",
    "ERROR": "CC0000",
}


# ─────────────────────────────────────────────────────────────────────────────
# ReportGenerator
# ─────────────────────────────────────────────────────────────────────────────

class ReportGenerator:
    """Collects test results and outputs multiple report formats."""

    def __init__(self, config: ReportConfig):
        self._cfg     = config
        self._results: List[TestResult] = []
        self._start   = time.time()
        Path(config.output_dir).mkdir(parents=True, exist_ok=True)

    # ── Add result ────────────────────────────────────────────────────────────

    def add_result(self, result: TestResult):
        self._results.append(result)

    def add_from_pytest_report(self, report, feature: str = "UNKNOWN",
                                asil: str = "QM",
                                req_ids: List[str] = None):
        """Ingest a pytest TestReport object."""
        if report.when != "call":
            return
        if report.passed:
            status = "PASS"
        elif report.failed:
            status = "FAIL"
        elif report.skipped:
            status = "SKIP"
        else:
            status = "ERROR"

        error_msg = ""
        if hasattr(report, "longreprtext"):
            error_msg = str(report.longreprtext)[:500]
        elif hasattr(report, "longrepr"):
            error_msg = str(report.longrepr)[:500]

        self.add_result(TestResult(
            test_id    = report.nodeid,
            feature    = feature,
            status     = status,
            duration_ms = getattr(report, "duration", 0.0) * 1000,
            asil       = asil,
            req_ids    = req_ids or [],
            error_msg  = error_msg,
        ))

    # ── Generate all ─────────────────────────────────────────────────────────

    def finalize(self):
        """Write all configured report formats."""
        cfg = self._cfg
        if cfg.html_report:
            path = self._write_html()
            log.info(f"HTML report: {path}")
        if cfg.excel_report and _OPENPYXL:
            path = self._write_excel()
            log.info(f"Excel report: {path}")
        if cfg.allure_results_dir:
            count = self._write_allure()
            log.info(f"Allure results: {count} files → {cfg.allure_results_dir}")
        self._log_summary()

    # ── HTML ──────────────────────────────────────────────────────────────────

    def _write_html(self) -> str:
        results = self._results
        total   = len(results)
        passed  = sum(1 for r in results if r.status == "PASS")
        failed  = sum(1 for r in results if r.status == "FAIL")
        skipped = sum(1 for r in results if r.status == "SKIP")
        errors  = sum(1 for r in results if r.status == "ERROR")
        duration_s = time.time() - self._start

        rows = ""
        for r in results:
            color    = _COLOR.get(r.status, "#888")
            req_str  = ", ".join(r.req_ids) or "-"
            err_clip = _html.escape(r.error_msg[:200]) if r.error_msg else ""
            rows += (
                f"<tr>"
                f"<td style='color:{color}'><b>{r.status}</b></td>"
                f"<td>{_html.escape(r.test_id)}</td>"
                f"<td>{r.feature}</td>"
                f"<td>{r.asil}</td>"
                f"<td>{r.duration_ms:.0f} ms</td>"
                f"<td>{req_str}</td>"
                f"<td style='font-size:11px;color:#999'>{err_clip}</td>"
                f"</tr>\n"
            )

        html_content = f"""<!DOCTYPE html>
<html lang="en"><head>
<meta charset="UTF-8">
<title>ADAS Test Report</title>
<style>
  body {{ font-family: 'Segoe UI', sans-serif; background:#1a1a2e; color:#eee; margin:20px; }}
  h1   {{ color:#00b4d8; }}
  .summary {{ display:flex; gap:20px; margin:20px 0; }}
  .card {{ padding:16px 24px; border-radius:8px; text-align:center; }}
  .pass  {{ background:#005f3c; }} .fail {{ background:#7d0000; }}
  .skip  {{ background:#7a5500; }} .total {{ background:#1f4e79; }}
  table {{ width:100%; border-collapse:collapse; }}
  th    {{ background:#16213e; padding:10px; text-align:left; color:#00b4d8; }}
  td    {{ padding:8px; border-bottom:1px solid #333; font-size:13px; }}
  tr:hover {{ background:#222; }}
</style>
</head><body>
<h1>🚗 ADAS Automated Test Report</h1>
<p>Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} | Duration: {duration_s:.1f}s</p>
<div class="summary">
  <div class="card total"><div style="font-size:2em">{total}</div><div>Total</div></div>
  <div class="card pass"><div style="font-size:2em">{passed}</div><div>Passed</div></div>
  <div class="card fail"><div style="font-size:2em">{failed}</div><div>Failed</div></div>
  <div class="card skip"><div style="font-size:2em">{skipped+errors}</div><div>Skip/Err</div></div>
</div>
<table>
<tr><th>Status</th><th>Test</th><th>Feature</th><th>ASIL</th><th>Duration</th><th>Requirements</th><th>Error</th></tr>
{rows}
</table>
</body></html>"""

        path = os.path.join(self._cfg.output_dir, "adas_test_report.html")
        with open(path, "w") as fh:
            fh.write(html_content)
        return path

    # ── Excel ─────────────────────────────────────────────────────────────────

    def _write_excel(self) -> str:
        wb = Workbook()

        # ── Summary sheet ──
        ws = wb.active
        ws.title = "Summary"
        headers = ["Status", "Test ID", "Feature", "ASIL", "Duration (ms)",
                   "Requirements", "Error Message"]
        ws.append(headers)
        for cell in ws[1]:
            cell.font      = Font(bold=True, color="FFFFFF")
            cell.fill      = PatternFill("solid", fgColor="16213E")
            cell.alignment = Alignment(horizontal="center")

        for r in self._results:
            row = [r.status, r.test_id, r.feature, r.asil,
                   round(r.duration_ms, 0), ", ".join(r.req_ids),
                   r.error_msg[:200]]
            ws.append(row)
            fill_color = _EXCEL_COLOR.get(r.status, "888888")
            ws.cell(ws.max_row, 1).fill = PatternFill("solid", fgColor=fill_color)

        # Auto-width
        for col in ws.columns:
            max_len = max(len(str(cell.value or "")) for cell in col)
            ws.column_dimensions[col[0].column_letter].width = min(max_len + 4, 80)

        # ── Per-feature sheets ──
        features: Dict[str, List[TestResult]] = {}
        for r in self._results:
            features.setdefault(r.feature, []).append(r)

        for feature, results in features.items():
            ws2 = wb.create_sheet(title=feature[:31])
            ws2.append(headers)
            for r in results:
                ws2.append([r.status, r.test_id, r.feature, r.asil,
                             round(r.duration_ms, 0), ", ".join(r.req_ids),
                             r.error_msg[:200]])
                fill_color = _EXCEL_COLOR.get(r.status, "888888")
                ws2.cell(ws2.max_row, 1).fill = PatternFill("solid", fgColor=fill_color)

        path = os.path.join(self._cfg.output_dir, "adas_test_report.xlsx")
        wb.save(path)
        return path

    # ── Allure JSON ───────────────────────────────────────────────────────────

    def _write_allure(self) -> int:
        out_dir = Path(self._cfg.allure_results_dir)
        out_dir.mkdir(parents=True, exist_ok=True)
        count = 0
        for r in self._results:
            allure_status = {
                "PASS": "passed", "FAIL": "failed",
                "SKIP": "skipped", "ERROR": "broken",
            }.get(r.status, "unknown")

            result_json = {
                "uuid":      f"{r.test_id.replace('/', '_').replace('::', '__')}_{int(r.timestamp)}",
                "name":      r.test_id,
                "status":    allure_status,
                "duration":  int(r.duration_ms),
                "start":     int(r.timestamp * 1000),
                "stop":      int((r.timestamp + r.duration_ms / 1000) * 1000),
                "labels": [
                    {"name": "feature",  "value": r.feature},
                    {"name": "severity", "value": r.asil},
                    {"name": "suite",    "value": r.feature},
                ],
                "links": [
                    {"name": req, "url": f"https://jira.example.com/browse/{req}", "type": "issue"}
                    for req in r.req_ids
                ],
                "statusDetails": {
                    "message": r.error_msg[:500] if r.error_msg else None,
                },
            }
            filename = out_dir / f"{result_json['uuid']}-result.json"
            with open(filename, "w") as fh:
                json.dump(result_json, fh, indent=2)
            count += 1
        return count

    # ── Summary ───────────────────────────────────────────────────────────────

    def _log_summary(self):
        total   = len(self._results)
        passed  = sum(1 for r in self._results if r.status == "PASS")
        failed  = sum(1 for r in self._results if r.status == "FAIL")
        rate    = f"{passed/total:.1%}" if total else "N/A"
        log.info(
            f"Report finalized | total={total} passed={passed} "
            f"failed={failed} pass_rate={rate}"
        )

    def summary(self) -> dict:
        total  = len(self._results)
        passed = sum(1 for r in self._results if r.status == "PASS")
        failed = sum(1 for r in self._results if r.status == "FAIL")
        return {
            "total": total,
            "passed": passed,
            "failed": failed,
            "pass_rate": round(passed / total, 4) if total else 0.0,
        }
